swarm/storage/mru/resource.go

```diff
@@ -153,693 +108,14 @@ func (r *resource) Size(_ chan bool) (int64, error) {
-func (h *Handler) get(rootAddr storage.Address) *resource {
-	if len(rootAddr) < storage.KeyLength {
-		log.Warn("Handler.get with invalid rootAddr")
-		return nil
-	}
-	hashKey := *(*uint64)(unsafe.Pointer(&rootAddr[0]))
-	h.resourceLock.RLock()
-	defer h.resourceLock.RUnlock()
-	rsrc := h.resources[hashKey]
-	return rsrc
-}
-
-// Sets the resource index value for the given nameHash
-func (h *Handler) set(rootAddr storage.Address, rsrc *resource) {
-	if len(rootAddr) < storage.KeyLength {
-		log.Warn("Handler.set with invalid rootAddr")
-		return
-	}
-	hashKey := *(*uint64)(unsafe.Pointer(&rootAddr[0]))
-	h.resourceLock.Lock()
-	defer h.resourceLock.Unlock()
-	h.resources[hashKey] = rsrc
-}
-
-// Checks if we already have an update on this resource, according to the value in the current state of the resource index
-func (h *Handler) hasUpdate(rootAddr storage.Address, period uint32) bool {
-	rsrc := h.get(rootAddr)
-	return rsrc != nil && rsrc.period == period
-}
-
-// create an update chunk. //TODO: Move to a method of SignedUpdateResource
-func newUpdateChunk(mru *SignedResourceUpdate) (*storage.Chunk, error) {
-
-	if len(mru.rootAddr) != storage.KeyLength || len(mru.metaHash) != storage.KeyLength {
-		log.Warn("Call to newUpdateChunk with nil rootAddr or metaHash")
-		return nil, NewError(ErrInvalidValue, "newUpdateChunk called without rootAddr or metaHash set")
-	}
-	// a datalength field set to 0 means the content is a multihash
-	var datalength int
-	if !mru.multihash {
-		datalength = len(mru.data)
-		if datalength == 0 {
-			return nil, NewError(ErrInvalidValue, "cannot update a resource with no data")
-		}
-	}
-
-	// prepend version, period, metaHash and rootAddr references
-	headerlength := 4 + 4 + storage.KeyLength + storage.KeyLength
-
-	actualdatalength := len(mru.data)
-	chunk := storage.NewChunk(mru.updateAddr, nil)
-	chunk.SData = make([]byte, 2+2+signatureLength+headerlength+actualdatalength) // initial 4 are uint16 length descriptors for headerlength and datalength
-
-	// data header length does NOT include the header length prefix bytes themselves
-	cursor := 0
-	binary.LittleEndian.PutUint16(chunk.SData[cursor:], uint16(headerlength))
-	cursor += 2
-
-	// data length
-	binary.LittleEndian.PutUint16(chunk.SData[cursor:], uint16(datalength))
-	cursor += 2
-
-	// header = period + version + rootAddr + metaHash
-	binary.LittleEndian.PutUint32(chunk.SData[cursor:], mru.period)
-	cursor += 4
-
-	binary.LittleEndian.PutUint32(chunk.SData[cursor:], mru.version)
-	cursor += 4
-
-	copy(chunk.SData[cursor:], mru.rootAddr[:storage.KeyLength])
-	cursor += storage.KeyLength
-	copy(chunk.SData[cursor:], mru.metaHash[:storage.KeyLength])
-	cursor += storage.KeyLength
-
-	// add the data
-	copy(chunk.SData[cursor:], mru.data)
-
-	// signature is the last item in the chunk data
-
-	cursor += actualdatalength
-	copy(chunk.SData[cursor:], mru.signature[:])
-
-	chunk.Size = int64(len(chunk.SData))
-	return chunk, nil
-}
-
```

Comment:
This is where the `Handler` moves to `handler.go`.
I made the mistake of not making a commit just for this, or maybe squashed at some point prior to one of the many rebases. Please accept my apologies.
The diff can be reviewed in detail here: [http://www.mergely.com/mI2Z3XZA/](http://www.mergely.com/mI2Z3XZA/)

Summary of changes:
* `parseUpdate` is now a method of `SignedResourceUpdate`, therefore refactor calls to it in `Handler`.
* `UpdateRequest` is renamed to a simpler name `Request`, so update references to it in `Handler`.
* `Handler.newMetaChunk` is refactored to a method of `resourceMetadata`, so update references to it in `Handler` and move that code to `metadata.go`
* `resourceData` is renamed back to `resourceUpdate`, and `resourceUpdate` is divided between header and data, so update references to it in `Handler`.
* Instantiation of  `LookupParams` variables are refactored to `LookupParams` factories whose name indicate what a particular combination of params actually does: Version=0, or Period=0 have special meanings and looked cryptic to me what the purpose of a particular initialization was.
* `Handler.lookup` is refactored to admit a `LookupParams` struct rather then the components of `LookupParams` separately.
* `resourceUpdateChunkAddr()` is refactored to `UpdateLookup.GetUpdateAddr()`. The new `UpdateLookup` structure represents the three components of an update search key: `period`, `version` and `rootAddr`. Thus, `UpdateLookup.GetUpdateAddr()` hashes them together to produce the lookup key.
`newUpdateChunk` is refactored to `SignedResourceUpdate.newUpdateChunk()`, so move code to `update.go` and update references to it in `Handler` code.

Reply:
I've gone through this and it looks ok, but I must admit it's rather hard to keep the overview even when breaking up in the series of PRs. I guess we won't get around an extra pass of thorough scrutiny on the last one.

Reply:
That's fine. At the end, it is the final status what matters and that it is a net improvement to the codebase. Sorry for all the fuss. The next steps are much easier.